src/conn/conn_compact.c

```diff
@@ -21,21 +21,269 @@ __compact_server_run_chk(WT_SESSION_IMPL *session)
+  WT_SESSION_IMPL *session, WT_BACKGROUND_COMPACT_STAT *compact_stat, uint64_t bucket)
+{
+    WT_CONNECTION_IMPL *conn;
+
+    conn = S2C(session);
+
+    TAILQ_REMOVE(&(conn)->background_compact.compacthash[bucket], compact_stat, hashq);
+    WT_ASSERT(session, (conn)->background_compact.file_count > 0);
+    --(conn)->background_compact.file_count;
+    WT_STAT_CONN_DECR(session, background_compact_files_tracked);
+
+    __wt_free(session, compact_stat->uri);
+    __wt_free(session, compact_stat);
+}
+
+/*
+ * __background_compact_get_stat --
+ *     Get the statistics for the given uri and id. The id ensures uniqueness in the event of
+ *     dropping and recreating files of the same name.
+ */
+static WT_BACKGROUND_COMPACT_STAT *
+__background_compact_get_stat(WT_SESSION_IMPL *session, const char *uri, int64_t id)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat, *temp_compact_stat;
+    WT_CONNECTION_IMPL *conn;
+    uint64_t bucket, hash;
+
+    conn = S2C(session);
+
+    WT_ASSERT(session, uri != NULL);
+
+    hash = __wt_hash_city64(uri, strlen(uri));
+    bucket = hash & (conn->hash_size - 1);
+
+    /* Find the uri in the files compacted list. */
+    TAILQ_FOREACH_SAFE(
+      compact_stat, &conn->background_compact.compacthash[bucket], hashq, temp_compact_stat)
+    if (strcmp(uri, compact_stat->uri) == 0) {
+        /*
+         * If we've found an entry in the list with the same URI but different ID's we must've
+         * dropped and recreated this table. Reset the entry in this case.
+         */
+        if (id != compact_stat->id) {
+            __background_compact_list_remove(session, compact_stat, bucket);
+            return (NULL);
+        }
+
+        return (compact_stat);
+    }
+
+    return (NULL);
+}
+
+/*
+ * __background_compact_should_run --
+ *     Check whether we should proceed with calling compaction on the given file.
+ */
+static bool
+__background_compact_should_run(WT_SESSION_IMPL *session, const char *uri, int64_t id)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat;
+    WT_CONNECTION_IMPL *conn;
+    uint64_t cur_time;
+
+    conn = S2C(session);
+
+    /* The history store file should not be compacted. */
+    if (WT_STREQ(uri, WT_HS_URI))
+        return (false);
+
+    /* If we haven't seen this file before we should try and compact it. */
+    compact_stat = __background_compact_get_stat(session, uri, id);
+    if (compact_stat == NULL)
+        return (true);
+
+    /* If we have been unsuccessful recently skip this file for some time. */
+    cur_time = __wt_clock(session);
+    if (WT_CLOCKDIFF_SEC(cur_time, compact_stat->prev_compact_time) >
+      conn->background_compact.max_file_skip_time)
+        return (true);
+
+    if (!compact_stat->prev_compact_success) {
+        compact_stat->skip_count++;
+        conn->background_compact.files_skipped++;
+        return (false);
+    }
+
+    /* If the last compaction pass was less successful than the average. Skip it for some time. */
```

Comment:
I don't think you want a  full stop between the two sentences. I think they make more sense when combined, what do you think?